### 1. train.csv 불러오기

In [1]:
import pandas as pd
pd.set_option("display.float_format", "{:.2f}".format)

df = pd.read_csv("train.csv")

### 2. 상위 5개 행 출력

In [2]:
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.00,1,0,A/5 21171,7.25,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.00,1,0,PC 17599,71.28,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.00,0,0,STON/O2. 3101282,7.92,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.00,1,0,113803,53.10,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.00,0,0,373450,8.05,NaN,S


### 3. 열 정보 확인


In [3]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    str    
 4   Sex          891 non-null    str    
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    str    
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    str    
 11  Embarked     889 non-null    str    
dtypes: float64(2), int64(5), str(5)
memory usage: 83.7 KB


**[분석] 결측치가 존재하는 열과 결측 개수**

전체 데이터는 891행이며, Non-Null Count가 891보다 작은 열이 결측치가 있는 열이다.

| 열 | 결측 개수 |
| --- | --- |
| Age | 177개 |
| Cabin | 687개 |
| Embarked | 2개 |

**[분석] dtype이 예상과 다르거나 주의가 필요한 열**

| 열 | dtype | 주의할 점 |
| --- | --- | --- |
| Age | float64 | 나이는 대부분 정수지만(0.42세 같은 소수 나이도 있음), 결측치(NaN)가 float 타입이라 열 전체가 float64로 저장됨 |
| Survived, Pclass | int64 | 숫자지만 계산용 수치가 아니라 생존 여부·객실 등급을 구분하는 범주형 값임 (Pclass는 1<2<3 순서는 있음) |
| Sex | str | 문자열이라 분석·계산에 쓰려면 숫자로 바꿔야 함 (12번에서 인코딩) |

### 4. Age, Fare의 평균·최솟값·최댓값

In [4]:
print("Age 평균:", df["Age"].mean().round(2))
print("Age 최댓값:", df["Age"].max().round(2))
print("Age 최솟값:", df["Age"].min().round(2))

print("Fare 평균:", df["Fare"].mean().round(2))
print("Fare 최댓값:", df["Fare"].max().round(2))
print("Fare 최솟값:", df["Fare"].min().round(2))

Age 평균: 29.7
Age 최댓값: 80.0
Age 최솟값: 0.42
Fare 평균: 32.2
Fare 최댓값: 512.33
Fare 최솟값: 0.0


### 5. 생존자·사망자 수

In [5]:
counts = df["Survived"].value_counts()

print(f"사망자(0): {counts[0]}명")
print(f"생존자(1): {counts[1]}명")

사망자(0): 549명
생존자(1): 342명


### 6. 객실 등급별 탑승객 수

In [6]:
counts = df["Pclass"].value_counts()
print(f"1등급: {counts[1]}명")
print(f"2등급: {counts[2]}명")
print(f"3등급: {counts[3]}명")

1등급: 216명
2등급: 184명
3등급: 491명


### 7. 50세 이상 탑승객 추출

In [7]:
df_50 = df[df["Age"]>= 50] 

df_50.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.00,0,0,17463,51.86,E46,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.00,0,0,113783,26.55,C103,S
15,16,1,2,"Hewlett, Mrs. (Mary D Kingcome)",female,55.00,0,0,248706,16.00,NaN,S
33,34,0,2,"Wheadon, Mr. Edward H",male,66.00,0,0,C.A. 24579,10.50,NaN,S
54,55,0,1,"Ostby, Mr. Engelhart Cornelius",male,65.00,0,1,113509,61.98,B30,C


### 8. 나이대(AgeGroup) 열 추가

In [8]:
def age_group(age):
    if pd.isna(age):
        return "미확인"
    elif age < 10:
        return "아동"
    elif age < 20:
        return "10대"
    elif age < 30:
        return "20대"
    elif age < 40:
        return "30대"
    elif age < 50:
        return "40대"
    elif age < 60:
        return "50대"
    else:
        return "60대 이상"

df["AgeGroup"]= df["Age"].apply(age_group)    

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,AgeGroup
0,1,0,3,"Braund, Mr. Owen Harris",male,22.00,1,0,A/5 21171,7.25,NaN,S,20대
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.00,1,0,PC 17599,71.28,C85,C,30대
2,3,1,3,"Heikkinen, Miss. Laina",female,26.00,0,0,STON/O2. 3101282,7.92,NaN,S,20대
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.00,1,0,113803,53.10,C123,S,30대
4,5,0,3,"Allen, Mr. William Henry",male,35.00,0,0,373450,8.05,NaN,S,30대


### 9. 성별·객실 등급별 평균 생존율

In [9]:
df.groupby(["Sex","Pclass"])["Survived"].mean().round(2)

Sex     Pclass
female  1        0.97
        2        0.92
        3        0.50
male    1        0.37
        2        0.16
        3        0.14
Name: Survived, dtype: float64

### 10. 나이대(AgeGroup)별 평균 생존율

In [10]:
df.groupby("AgeGroup")["Survived"].mean().round(2)

AgeGroup
10대      0.40
20대      0.35
30대      0.44
40대      0.38
50대      0.42
60대 이상   0.27
미확인      0.29
아동       0.61
Name: Survived, dtype: float64

### 11. 열별 결측치 개수와 비율 (내림차순)

In [11]:
missing = df.isna().sum()

ratio= (missing / len(df) * 100).round(2) 

result = pd.DataFrame({
    "결측치 수": missing,
    "비율(%)": ratio
})

result.sort_values("결측치 수",ascending=False)

,결측치 수,비율(%)
Cabin,687,77.10
Age,177,19.87
Embarked,2,0.22
Survived,0,0.00
PassengerId,0,0.00
Sex,0,0.00
Name,0,0.00
Pclass,0,0.00
SibSp,0,0.00
Ticket,0,0.00


### 12. 성별 인코딩 (Gender_Encoded 열 추가)

In [12]:
df["Gender_Encoded"] = df["Sex"].map({"male":0,"female":1})

df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked,AgeGroup,Gender_Encoded
0,1,0,3,"Braund, Mr. Owen Harris",male,22.00,1,0,A/5 21171,7.25,NaN,S,20대,0
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.00,1,0,PC 17599,71.28,C85,C,30대,1
2,3,1,3,"Heikkinen, Miss. Laina",female,26.00,0,0,STON/O2. 3101282,7.92,NaN,S,20대,1
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.00,1,0,113803,53.10,C123,S,30대,1
4,5,0,3,"Allen, Mr. William Henry",male,35.00,0,0,373450,8.05,NaN,S,30대,0


### 13. 탑승지(Embarked)별 평균 요금

In [13]:
df.groupby("Embarked")["Fare"].mean().round(2)

Embarked
C   59.95
Q   13.28
S   27.08
Name: Fare, dtype: float64

### 14. 피벗 테이블 (Pclass × Sex, Fare 평균)

In [14]:
df.pivot_table(index="Pclass",columns="Sex",values="Fare",aggfunc="mean").round(2)

Sex,female,male
Pclass,,
1,106.13,67.23
2,21.97,19.74
3,16.12,12.66


### 15. 가족 수(FamilySize) 열 추가 및 요약 통계

In [15]:
df["FamilySize"] = df["SibSp"]+ df["Parch"]

df["FamilySize"].describe().round(2)

count   891.00
mean      0.90
std       1.61
min       0.00
25%       0.00
50%       0.00
75%       1.00
max      10.00
Name: FamilySize, dtype: float64

### 16. 이름에서 호칭(Title) 추출

In [16]:
df["Title"] = df["Name"].str.extract(r', ([A-Za-z]+)\.')

df["Title"].value_counts().head(5)

Title
Mr        517
Miss      182
Mrs       125
Master     40
Dr          7
Name: count, dtype: int64

### 17. 호칭(Title)별 승객 수·평균 나이·평균 생존율

In [17]:
df.groupby("Title").agg(
    승객수=("PassengerId","count"),
    평균나이=("Age","mean"),
    평균생존율=("Survived","mean")
).round(2)

,승객수,평균나이,평균생존율
Title,,,
Capt,1,70.00,0.00
Col,2,58.00,0.50
Don,1,40.00,0.00
Dr,7,42.00,0.43
Jonkheer,1,38.00,0.00
Lady,1,48.00,1.00
Major,2,48.50,0.50
Master,40,4.57,0.57
Miss,182,21.77,0.70


### 18. 생존자·사망자 나이 분포 비교 (히스토그램)

In [18]:
survived = df[df["Survived"]==1]["Age"]
died = df[df["Survived"]==0]["Age"]

import matplotlib.pyplot as plt

plt.rcParams["font.family"] = "Malgun Gothic"
plt.rcParams["axes.unicode_minus"] = False

fig, ax = plt.subplots()

ax.hist(survived.dropna(), bins=20, alpha=0.5, label="생존")
ax.hist(died.dropna(), bins=20, alpha=0.5, label="사망")

ax.set_title("생존자, 사망자 나이 분포")
ax.set_xlabel("나이")
ax.set_ylabel("승객 수")
ax.legend()

fig.savefig("age_distribution.png")
plt.close()

### 19. Title·Pclass 그룹별 나이 중앙값으로 Age 결측치 채우기

In [19]:
median_age =df.groupby(["Title","Pclass"])["Age"].transform("median")

df["Age"] = df["Age"].fillna(median_age)

df["Age"].isna().sum()

np.int64(0)

### 20. 수치형 변수 상관관계 히트맵

In [20]:
cols = ["Survived", "Pclass", "Age", "SibSp", "Parch", "Fare"]

corr = df[cols].corr().round(2)

corr

import seaborn as sns

fig, ax = plt.subplots(figsize=(8, 6))

sns.heatmap(corr, annot=True, fmt=".2f", cmap="coolwarm", center=0, ax=ax)

ax.set_title("수치형 변수 상관관계 히트맵")

fig.savefig("heatmap.png")
plt.close()